proposals/0010-resource-element-type-validation.md

```diff
@@ -0,0 +1,221 @@
+* Proposal: [0010](0010-resource-element-type-validation.md)
+* Author(s): [Joshua Batista](https://github.com/bob80905)
+* Sponsor: Joshua Batista
+* Status: **Under Consideration**
+* Impacted Project(s): (LLVM)
+
+* Issues: [#75676](https://github.com/llvm/llvm-project/issues/75676)
+
+## Introduction
+Resources are often used in HLSL, with various resource element types (RETs).
+For example:
+```
+RWBuffer<float> rwbuf: register(u0);
+```
+In this code, the RET is `float`, and the resource type is `RWBuffer`.
+All resources can be placed in two categories, raw buffers and typed buffers. 
+Below is a description of all resources and their corresponding categories
+* raw buffers
+  * [Append|Consume|RW]StructuredBuffer
+  * [RW]ByteAddressBuffer
+* typed buffers
+  * [RW]Buffer
+  * [Feedback]Texture*
+
+There is a distinct set of rules that define valid RETs for raw buffer resources, 
+and a separate set of rules that define valid RETs for typed buffer resources.
+
+RETs for typed buffer resources:
+* Are not intangible (e.g., isn't a resource type)
+* Must be vectors or scalars of arithmetic types, not bools nor enums nor arrays
+* The type should be line-vector layout compatible (homogenous, at most 4 elements, and at most 128 bits in size) 
+
+RETs for raw buffer variants are much less constrained:
+* Are not intangible (e.g., isn't a resource type)
+* All constituent types must be arithmetic types or bools or enums
+
+Resource types are never allowed as RETs (i.e., `RWBuffer<int>` as an RET).
+If someone writes `RWBuffer<MyCustomType>` and MyCustomType is not a 
+valid RET, there should be infrastructure to reject this RET and emit a message 
+explaining why it was rejected as an RET.
+
+## Motivation
+Currently, there is an allow list of valid RETs. It must be modified with respect 
+to this spec. Anything that is not a valid RET will be rejected. The allow list isn't
+broad enough, because there is no case where the RET is user-defined for a raw buffer.
+Ideally, a user should be able to determine how any user-defined structure is invalid 
+as an RET. Some system should be in place to more completely enforce the rules for 
+valid and invalid RETs, as well as provide useful information on why they are invalid.
+
+For example, when targeting DXIL IR, `RWBuffer<double4> b : register(u4);` will emit
+an error in DXC, but will not in clang-dxc, despite the fact that `double4` is an 
+invalid RET for typed buffers.
+
+## Proposed solution
+
+The proposed solution is to modify the declaration of each resource declared in 
+`clang\lib\Sema\HLSLExternalSemaSource.cpp` and insert into each representative
+AST node a concept. The AST node will be created as if the C++20 `concept` keyword
+was parsed and applied to the declaration. The concept will be used to validate the
+given RET, and will emit errors when the given RET is invalid.

```

Comment:
Wonder if something along these lines would help capture some of the discussions that led up to this decision?

> Requirements such as this are well expressed using the C++20 concepts. These aren't currently supported
> in HLSL, but we expect support to be added at some point in the future. However, LLVM does support
> concepts and it is possible to make use of them even if they can't be spelled in HLSL source code. The way
> to do this is to construct the appropriate AST nodes in Sema. In fact, this is already done for other
> non-spellable aspects of resources in `HLSLExternalSemaSource.cpp`, and so we have a natural place to
> add these concepts.